src/RingBuffers.jl

```diff
@@ -12,8 +12,43 @@ import Base: unsafe_convert, pointer
 import Base: isopen, close
 
 using Base: AsyncCondition
+import Compat
+import Compat: Libdl, Cvoid, undef, popfirst!, @compat
+
+function init_pa_ringbuffer()
+    libdir = joinpath(dirname(@__FILE__), "..", "deps", "usr", "lib")
+    libsuffix = ""
+    @static if Compat.Sys.islinux() && Sys.ARCH == :x86_64
+        libsuffix = "x86_64-linux-gnu"
+    elseif Compat.Sys.islinux() && Sys.ARCH == :i686
+        libsuffix = "i686-linux-gnu"
+    elseif Compat.Sys.isapple() && Sys.ARCH == :x86_64
+        libsuffix = "x86_64-apple-darwin14"
+    elseif Compat.Sys.iswindows() && Sys.ARCH == :x86_64
+        libsuffix = "x86_64-w64-mingw32"
+    elseif Compat.Sys.iswindows() && Sys.ARCH == :i686
+        libsuffix = "i686-w64-mingw32"
+    elseif !any(
+            (sfx) -> isfile(joinpath(libdir, "pa_ringbuffer.$sfx")),
+            ("so", "dll", "dylib"))
+        error("Unsupported platform $(Sys.MACHINE). You can build your own library by running `make` from $(joinpath(@__FILE__, "..", "deps", "src"))")
+    end
+    # if there's a suffix-less library, it was built natively on this machine,
+    # so load that one first, otherwise load the pre-built one
+    libpa_ringbuffer = Libdl.find_library(
+            ["pa_ringbuffer", "pa_ringbuffer_$libsuffix"],
+            [libdir])
+    libpa_ringbuffer == "" && error("Could not load pa_ringbuffer library, please file an issue at https://github.com/JuliaAudio/RingBuffers.jl/issues with your `versioninfo()` output")
+    # override dlopen flags to make sure we always use `RTLD_GLOBAL` so that the
+    # library functions are available to other C shim libraries that other
+    # packages might need to add to handle their audio callbacks.
+    Libdl.dlopen(libpa_ringbuffer, Libdl.RTLD_LAZY |
+                                   Libdl.RTLD_DEEPBIND |
+                                   Libdl.RTLD_GLOBAL)
+    return libpa_ringbuffer
+end
 
-__init__() = init_pa_ringbuffer()
+const libpa_ringbuffer = init_pa_ringbuffer()
```

Comment:
Can you give a little more context for this change? Previously it would wait until runtime to use `dlopen` to find the library. If I understand correctly doing it this way will run `init_pa_ringbuffer()` at precompile time, so I'd worry that future runs would have a messed up pointer to the library.

Reply:
Hi, I think that I was wrong. and I'm going to use BinaryProvider to generate the `deps.jl` file during at the build time, that's probably more common for Julia packages.